File: src/main/java/com/lessmarkup/userinterface/nodehandlers/user/ResetPasswordNodeHandler.java
package com.lessmarkup.userinterface.nodehandlers.user;

import com.google.inject.Inject;
import com.lessmarkup.Constants;
import com.lessmarkup.TextIds;
import com.lessmarkup.dataobjects.User;
import com.lessmarkup.framework.helpers.DependencyResolver;
import com.lessmarkup.framework.helpers.LanguageHelper;
import com.lessmarkup.framework.helpers.LoggingHelper;
import com.lessmarkup.framework.system.RequestContextHolder;
import com.lessmarkup.interfaces.cache.DataCache;
import com.lessmarkup.interfaces.data.DomainModel;
import com.lessmarkup.interfaces.data.DomainModelProvider;
import com.lessmarkup.interfaces.security.UserSecurity;
import com.lessmarkup.interfaces.system.RequestContext;
import com.lessmarkup.userinterface.model.user.ChangePasswordModel;
import com.lessmarkup.userinterface.nodehandlers.common.DialogNodeHandler;
import scala.Option;

import java.io.IOException;
import java.time.OffsetDateTime;

public class ResetPasswordNodeHandler extends DialogNodeHandler<ChangePasswordModel> {

    private String ticket;
    private String email;

    private final DomainModelProvider domainModelProvider;
    private final UserSecurity userSecurity;

    @Inject
    public ResetPasswordNodeHandler(DataCache dataCache, DomainModelProvider domainModelProvider, UserSecurity userSecurity) {
        super(dataCache, ChangePasswordModel.class);
        this.domainModelProvider = domainModelProvider;
        this.userSecurity = userSecurity;
    }

    @Override
    protected ChangePasswordModel loadObject() {
        return DependencyResolver.resolve(ChangePasswordModel.class);
    }

    @Override
    protected String saveObject(ChangePasswordModel changedObject) {
        Option<Object> userId = userSecurity.validatePasswordChangeToken(email, ticket);

        if (!userId.isDefined()) {
            LoggingHelper.getLogger(getClass()).info("Cannot change password: cannot get valid user id");
            return LanguageHelper.getText(Constants.ModuleTypeMain(), TextIds.PASSWORD_CHANGE_ERROR);
        }

        try (DomainModel domainModel = domainModelProvider.create()) {
            User user = domainModel.query().findOrDefaultJava(User.class, (Long)userId.get());

            if (user == null) {
                LoggingHelper.getLogger(getClass()).info("Cannot change password: user id=" + userId.get() + " does not exist");
                return LanguageHelper.getText(Constants.ModuleTypeMain(), TextIds.PASSWORD_CHANGE_ERROR);
            }

            scala.Tuple2<String, String> result = userSecurity.changePassword(changedObject.getPassword());

            user.setPasswordChangeToken(null);
            user.setPasswordChangeTokenExpires(null);
            user.setPassword(result._2());
            user.setEmailConfirmed(true);
            user.setSalt(result._1());
            user.setLastPasswordChanged(OffsetDateTime.now());
            user.setEmailConfirmed(true);

            domainModel.update(user);

            return LanguageHelper.getText(Constants.ModuleTypeMain(), TextIds.PASSWORD_CHANGED);
        }
    }

    public void initialize(String email, String ticket) {
        this.email = email;
        this.ticket = ticket;
    }

    @Override
    public boolean trySubmitResponse(String path) throws IOException {
        if (path != null) {
            return false;
        }

        Option<Object> userId = userSecurity.validatePasswordChangeToken(email, ticket);

        RequestContext requestContext = RequestContextHolder.getContext();

        if (!userId.isDefined()) {
            requestContext.sendError(404);
            return true;
        }

        return false;
    }
}

File: src/main/scala/com/lessmarkup/userinterface/nodehandlers/user/ResetPasswordNodeHandler.scala
package com.lessmarkup.userinterface.nodehandlers.user

import java.time.OffsetDateTime

import com.lessmarkup.dataobjects.User
import com.lessmarkup.framework.helpers.{DependencyResolver, LanguageHelper, LoggingHelper}
import com.lessmarkup.framework.nodehandlers.NodeHandlerConfiguration
import com.lessmarkup.framework.system.RequestContextHolder
import com.lessmarkup.interfaces.cache.DataCache
import com.lessmarkup.interfaces.data.{DomainModel, DomainModelProvider}
import com.lessmarkup.interfaces.security.UserSecurity
import com.lessmarkup.interfaces.structure.{NodeHandler, NodeHandlerFactory}
import com.lessmarkup.interfaces.system.RequestContext
import com.lessmarkup.userinterface.model.user.ChangePasswordModel
import com.lessmarkup.userinterface.nodehandlers.common.DialogNodeHandler
import com.lessmarkup.{Constants, TextIds}

class ResetPasswordNodeHandlerFactory(dataCache: DataCache, domainModelProvider: DomainModelProvider, userSecurity: UserSecurity)
  extends NodeHandlerFactory {
  override def createNodeHandler(configuration: NodeHandlerConfiguration, arguments: Any*): NodeHandler = {
    if (arguments.length != 2) {
      throw new IllegalArgumentException
    }
    new ResetPasswordNodeHandler(dataCache, domainModelProvider, userSecurity, configuration, arguments(0).asInstanceOf, arguments(1).asInstanceOf)
  }
}

class ResetPasswordNodeHandler(
    dataCache: DataCache,
    domainModelProvider: DomainModelProvider,
    userSecurity: UserSecurity,
    configuration: NodeHandlerConfiguration,
    email: String,
    ticket: String
  )
  extends DialogNodeHandler[ChangePasswordModel](dataCache, classOf[ChangePasswordModel], configuration) {

  protected def loadObject: Option[ChangePasswordModel] = {
    Option(DependencyResolver.resolve(classOf[ChangePasswordModel]))
  }

  protected def saveObject(changedObject: Option[ChangePasswordModel]): String = {

    val userId = userSecurity.validatePasswordChangeToken(email, ticket)

    if (userId.isEmpty) {
      LoggingHelper.getLogger(getClass).info("Cannot change password: cannot get valid user id")
      return LanguageHelper.getText(Constants.ModuleTypeMain, TextIds.PASSWORD_CHANGE_ERROR)
    }

    val domainModel: DomainModel = domainModelProvider.create
    try {
      val userOption = domainModel.query.find(classOf[User], userId.get)
      if (userOption.isEmpty) {
        LoggingHelper.getLogger(getClass).info("Cannot change password: user id=" + userId.get + " does not exist")
        return LanguageHelper.getText(Constants.ModuleTypeMain, TextIds.PASSWORD_CHANGE_ERROR)
      }
      val user = userOption.get
      val result: (String, String) = userSecurity.changePassword(changedObject.get.getPassword)
      user.passwordChangeToken = None
      user.passwordChangeTokenExpires = None
      user.password = result._2
      user.emailConfirmed = true
      user.salt = result._1
      user.lastPasswordChanged = OffsetDateTime.now
      user.emailConfirmed = true
      domainModel.update(user)
      LanguageHelper.getText(Constants.ModuleTypeMain, TextIds.PASSWORD_CHANGED)
    } finally {
      if (domainModel != null) domainModel.close()
    }
  }

  override def trySubmitResponse(path: String): Boolean = {
    if (path != null) {
      return false
    }
    val userId = userSecurity.validatePasswordChangeToken(email, ticket)
    val requestContext: RequestContext = RequestContextHolder.getContext
    if (userId.isEmpty) {
      requestContext.sendError(404)
      return true
    }
    false
  }
}